nixos/qemu-vm: refactor bootDisk using make-disk-image

RaitoBezarius opened the pull request:
###### Description of changes

Depends on #207038.

A small cover letter on the changes:

We remove the `bootDisk` ad-hoc generation and rewire it up with `make-disk-image`, this unlocks various features:

- perform bootloader operations that have an impact on EFI variables (SecureBoot, etc.)
- factorization (everything is controlled by `make-disk-image`)
- clarification of the "current" boot configuration: https://github.com/NixOS/nixpkgs/pull/207039/files#diff-5f468553a33ed484ce0a34ef2f404fb7f37bd981ba95fa5ec666bcbdaeb3d6d9R253-R263 -- we have various mode of operations and this module is becoming confusing on what bunch of booleans realizes a specific configuration, pushing for an explicit string makes the code more maintainable and readable.
- exposes `rootDevice`, `bootDevice` and `bootPartition` which clarify further the different roles of the disk in the VM world we use, to cater for legacy (disk-level) and UEFI (partition-level) needs.
- undo `persistBootDevice` introduced in 678eed323ffd90117472cd432ebe85dddaff07f1

Testing is done on https://hydra.nixos.org/jobset/nixos/python-test-refactoring on a recent enough: nixos-unstable-small and treewide cleanups are done.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

RaitoBezarius committed:
nixos/tests(treewide): vdb → vda, vdc → vdb, vdd → vdc, … / bootDevice → rootDevice

RaitoBezarius committed:
nixos/qemu-vm: refactor bootDisk generation using make-disk-image

RaitoBezarius commented on nixos/lib/make-disk-image.nix:
```diff
@@ -303,6 +303,7 @@ let format' = format; in let
     [ rsync
       util-linux
       parted
+      gptfdisk
```
don't know why it didn't kept the other changes, anyway, will undo those

RaitoBezarius commented on nixos/maintainers/scripts/ec2/amazon-image.nix:
```diff
@@ -43,7 +43,7 @@ in {
     sizeMB = mkOption {
       type = with types; either (enum [ "auto" ]) int;
-      default = 2048;
+      default = 2252;
```
need to say why this is needed but basically: our closure size increased and unsure this is working on master at all

RaitoBezarius commented on nixos/modules/virtualisation/qemu-vm.nix:
```diff
@@ -361,10 +393,55 @@ in
+          '';
+      };
+
+    virtualisation.rootDevice =
```
this option may not be totally "finished"

blitz commented:
@RaitoBezarius The comments sound like you still want to work on this. What's the status here?

RaitoBezarius commented:
> @RaitoBezarius The comments sound like you still want to work on this. What's the status here?

Didn't have time to fix my own comments yet, but it's open to additional comments.

RaitoBezarius commented:
Re-evaluating this over Hydra to find out about potential failures, then I will address all remaining comments and we will be ready for another review phase.

RaitoBezarius commented:
According to https://hydra.nixos.org/eval/1792553, only 6 failing new jobs compared to nixos-unstable-small.

RaitoBezarius commented:
Sorted out:

- Amazon failure seems related to OOM
- Bootspec failure seems unrelated
- LUKS failing
- ZFS failing

I am working on fixing ZFS as it requires to disentangle the new "persistBootDevice" / tmpfs root device API.

RaitoBezarius commented:
Mostly everything fixed on x86_64-linux except a crash in a Proxmox test (https://hydra.nixos.org/build/213807775/nixlog/11/tail) and aarch64-linux tests are not in super good shape.

RaitoBezarius committed:
nixos/tests/bootspec: assert for boot.json rather than bootspec/boot.json

RaitoBezarius committed:
nixos/virtualisation/qemu-vm: remove persistBootDevice option

This option has been introduced in 678eed323ffd90117472cd432ebe85dddaff07f1 without realizing there was this
PR inflight, unfortunately, it collide with what this PR does and make
it irrelevant.

Therefore, I remove it here.

RaitoBezarius committed:
CODEOWNERS: add raitobezarius to qemu-vm.nix

I do a lot of work on QEMU VM and make-disk-image and I was bitten by an
unnotified change recently, so I want to chime in the future changes of
this file.

RaitoBezarius commented:
cc @pwaller as I am undoing a part of your commit.

pwaller commented:
> cc @pwaller as I am undoing a part of your commit.

Thanks for the ping. `persistBootDevice` was only added to support being able to test that changing initrd secrets worked:
https://github.com/NixOS/nixpkgs/pull/209156/files#diff-f55c5820b08601bdf7bd294c16d34ca5fe2f8cfa304e5a2f8f55787fb6fb5d99

If this test is still passing (it uses `grub-reboot` to select the boot entry, which I presume is persisted somehow in your changes) then I don't see a problem.

RaitoBezarius commented:
> > cc @pwaller as I am undoing a part of your commit.
> 
> Thanks for the ping. `persistBootDevice` was only added to support being able to test that changing initrd secrets worked: https://github.com/NixOS/nixpkgs/pull/209156/files#diff-f55c5820b08601bdf7bd294c16d34ca5fe2f8cfa304e5a2f8f55787fb6fb5d99
> 
> If this test is still passing (it uses `grub-reboot` to select the boot entry, which I presume is persisted somehow in your changes) then I don't see a problem.

Indeed, that's the case now, it is still passing.

RaitoBezarius commented:
Latest evaluation: https://hydra.nixos.org/eval/1793584?compare=-172800&full=0#tabs-now-fail
From there, it seems like there are only few problematic jobs remaining.

lilyinstarlight commented on nixos/modules/virtualisation/qemu-vm.nix:
```diff
@@ -365,22 +326,45 @@ in
     virtualisation.bootDevice =
       mkOption {
         type = types.path;
+        default = lookupDriveDeviceName "root" cfg.qemu.drives;
+        defaultText = literalExpression ''lookupDriveDeviceName "root" cfg.qemu.drives'';
         example = "/dev/vda";
         description =
           lib.mdDoc ''
-            The disk to be used for the root filesystem.
+            The disk to be used for the boot filesystem.
+            By default, it is the same disk as the root filesystem.
+          '';
+        };
```
Given this changes the meaning of the option (which it was sorta poorly named to begin with), I feel like we should at a minimum include release notes describing the change and addition of `virtualisation.rootDevice`

I also wonder if we should still rename the option anyway so we can add `mkRemovedOption` and give a useful error for any downstream users of it (I don't imagine there are many downstream users though, if any at all, so I do not feel strnogly on it)

RaitoBezarius replied:
We could rename the option but do you have a suggestion on the rename?

lilyinstarlight replied:
Maybe `bootLoaderDevice` or `bootDiskDevice`? Idk really, and I wish the `bootDevice` option hadn't already existed before this PR and meant "root device" (but oh well I guess)

RaitoBezarius replied:
`bootLoaderDevice` sounds good because it's really a "bootloader device" and lead to probably do a refactor `useBootLoader` -> `bootloader.{enable,device,enableEFI}`, etc. Thank you!

RaitoBezarius replied:
Applied!

lilyinstarlight commented on nixos/modules/virtualisation/qemu-vm.nix:
```diff
@@ -365,22 +326,45 @@ in
+      mkOption {
+        type = types.nullOr types.path;
+        default = if cfg.useEFIBoot then "${cfg.bootDevice}1" else null;
+        defaultText = literalExpression ''if cfg.useEFIBoot then "''${cfg.bootDevice}1" else null;'';
```
```suggestion
        defaultText = literalExpression ''if cfg.useEFIBoot then "''${cfg.bootDevice}1" else null'';
```

No semicolons in `defaultText`

RaitoBezarius replied:
Applied

lilyinstarlight commented on nixos/modules/virtualisation/qemu-vm.nix:
```diff
@@ -749,10 +733,22 @@ in
+        }).fd;
+        defaultText = ''(pkgs.OVMF.override {
+          secureBoot = cfg.useSecureBoot;
+        }).fd;'';
```
```suggestion
        }).fd'';
```

Same as above

RaitoBezarius replied:
Applied

lilyinstarlight reviewed (commented):
I'm not the most knowledgeable about the qemu-vm test stuff, but this change feels sane to me

Since it changes some semantics it could have some fallout for downstream users, but I don't imagine there are many and it should hopefully be easy for them to understand and update

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/2090

lilyinstarlight commented:
It looks like `nixosTests.bootspec.grub` is failing with this PR with this log snippet:

```
nixos-disk-image> installing the GRUB 2 boot loader on /dev/vda...
nixos-disk-image> Installing for i386-pc platform.
nixos-disk-image> /nix/store/07bws5lfvgz0a6psxbyhzzz0sbrnqdab-grub-2.06/sbin/grub-install: warning: this GPT partition label contains no BIOS Boot Partition; embedding won't be possible.
nixos-disk-image> /nix/store/07bws5lfvgz0a6psxbyhzzz0sbrnqdab-grub-2.06/sbin/grub-install: warning: Embedding is not possible.  GRUB can only be installed in this setup by using blocklists.  However, blocklists are UNRELIABLE and their use is discouraged..
nixos-disk-image> /nix/store/07bws5lfvgz0a6psxbyhzzz0sbrnqdab-grub-2.06/sbin/grub-install: error: will not proceed with blocklists.
nixos-disk-image> /nix/store/nj03jkzbac12apslfmhhmss8qrlvx7wg-install-grub.pl: installation of GRUB on /dev/vda failed: No such file or directory
```

RaitoBezarius commented:
> It looks like `nixosTests.bootspec.grub` is failing with this PR with this log snippet:
> 
> ```
> nixos-disk-image> installing the GRUB 2 boot loader on /dev/vda...
> nixos-disk-image> Installing for i386-pc platform.
> nixos-disk-image> /nix/store/07bws5lfvgz0a6psxbyhzzz0sbrnqdab-grub-2.06/sbin/grub-install: warning: this GPT partition label contains no BIOS Boot Partition; embedding won't be possible.
> nixos-disk-image> /nix/store/07bws5lfvgz0a6psxbyhzzz0sbrnqdab-grub-2.06/sbin/grub-install: warning: Embedding is not possible.  GRUB can only be installed in this setup by using blocklists.  However, blocklists are UNRELIABLE and their use is discouraged..
> nixos-disk-image> /nix/store/07bws5lfvgz0a6psxbyhzzz0sbrnqdab-grub-2.06/sbin/grub-install: error: will not proceed with blocklists.
> nixos-disk-image> /nix/store/nj03jkzbac12apslfmhhmss8qrlvx7wg-install-grub.pl: installation of GRUB on /dev/vda failed: No such file or directory
> ```

If I remember well, the failure is alas unrelated. It's because GRUB tries to install legacy for a UEFI only target.
I implemented the fix somewhere and lost it :D.

lilyinstarlight commented:
> If I remember well, the failure is alas unrelated. It's because GRUB tries to install legacy for a UEFI only target.

Yeah the error message sounded like that was the issue

It looks like the NixOS GRUB module assumes that it needs to install legacy unless `boot.loader.grub.device = "nodev"` is set (in which case if I'm following this Perl right (ugh) then it will assume it needs to do an EFI-only install)

RaitoBezarius commented:
> > If I remember well, the failure is alas unrelated. It's because GRUB tries to install legacy for a UEFI only target.
> 
> Yeah the error message sounded like that was the issue
> 
> It looks like the NixOS GRUB module assumes that it needs to install legacy unless `boot.loader.grub.device = "nodev"` is set (in which case if I'm following this Perl right (ugh) then it will assume it needs to do an EFI-only install)

Thank you for going in the rabbit hole, I rediscover this each time. :)

lilyinstarlight commented:
> Thank you for going in the rabbit hole, I rediscover this each time. :)

So I don't know if this is "correct" necessarily, but this diff on top of this PR does allow a number of grub tests to pass again (those which currently pass on master and fail on this PR):

```diff
diff --git a/nixos/modules/virtualisation/qemu-vm.nix b/nixos/modules/virtualisation/qemu-vm.nix
index f85567c42f2..fa928a1ba2f 100644
--- a/nixos/modules/virtualisation/qemu-vm.nix
+++ b/nixos/modules/virtualisation/qemu-vm.nix
@@ -849,7 +849,7 @@ in
             ${opt.writableStore} = false;
         '';
 
-    boot.loader.grub.device = mkVMOverride cfg.bootLoaderDevice;
+    boot.loader.grub.device = mkVMOverride (if cfg.useEFIBoot then "nodev" else cfg.bootLoaderDevice);
     boot.loader.grub.gfxmodeBios = with cfg.resolution; "${toString x}x${toString y}";
     virtualisation.rootDevice = mkDefault suggestedRootDevice;
 
```

RaitoBezarius commented:
@lilyinstarlight Adressed and latest evaluation shows that it fixes the GRUB tests: https://hydra.nixos.org/eval/1793709

nikstur commented on nixos/modules/virtualisation/qemu-vm.nix:
```diff
@@ -362,25 +324,48 @@ in
       mkOption {
-        type = types.bool;
-        default = false;
+        type = types.path;
```
Because it can also be `null` this should be an option type.

RaitoBezarius replied:
A null device is an error to force the NixOS module system to warn the user they did not set an explicit root device in "custom" mode, so working as intended?

nikstur replied:
Ah ok understood.

nikstur commented on nixos/modules/virtualisation/qemu-vm.nix:
```diff
@@ -297,17 +238,38 @@ let
     copyChannel = false;
   };
 
+  bootConfiguration =
+    if cfg.useDefaultFilesystems
+    then
+      if cfg.useBootLoader
+      then
+        if cfg.useEFIBoot then "efi_bootloading_with_default_fs"
+        else "legacy_bootloading_with_default_fs"
+      else
+        "direct_boot_with_default_fs"
+    else
+      "custom";
+  suggestedRootDevice = {
+    "efi_bootloading_with_default_fs" = "${cfg.bootLoaderDevice}2";
+    "legacy_bootloading_with_default_fs" = "${cfg.bootLoaderDevice}1";
+    "direct_boot_with_default_fs" = cfg.bootLoaderDevice;
+    "custom" = null;
+  }.${bootConfiguration};
```
I really really really like this. This brings this module so much clarity. Much more than documentation could have.

RaitoBezarius replied:
Thank you <3

nikstur commented on nixos/modules/virtualisation/qemu-vm.nix:
```diff
@@ -1087,8 +1055,8 @@ in
           neededForBoot = true;
         };
         # see note [Disk layout with `useBootLoader`]
```
This note doesn't exist anymore.

nikstur commented on nixos/modules/virtualisation/qemu-vm.nix:
```diff
@@ -200,95 +215,21 @@ let
+    format = "qcow2";
+    onlyNixStore = false;
+    partitionTableType = selectPartitionTableLayout { inherit (cfg) useDefaultFilesystems useEFIBoot; };
+    installBootLoader = cfg.useBootLoader && cfg.useDefaultFilesystems;
```
Why is the condition AND `cfg.useDefaultFilesystems`? This might deserve a comment.

RaitoBezarius replied:
Commented

nikstur reviewed (approved):
Very cool change. The comments and descriptions could use some more editing. Most of my comments are either nitpicks or revolve around naming. Although I think naming is very important, I think it is more important to get this PR across the finish line. We can still improve the naming later.

RaitoBezarius commented on nixos/modules/virtualisation/qemu-vm.nix:
```diff
@@ -297,17 +238,38 @@ let
     (mkRenamedOptionModule [ "virtualisation" "pathsInNixDB" ] [ "virtualisation" "additionalPaths" ])
+    (mkRemovedOptionModule [ "virtualisation" "bootDevice" ] "This option was renamed to `virtualisation.rootDevice`, as it was incorrectly named and misleading. Take the time to review what you want to do and look at the new options like `virtualisation.{bootLoaderDevice, bootPartition}`, open an issue in case of issues.")
+    (mkRemovedOptionModule [ "virtualisation" "efiVars" ] "This option was removed, it is possible to provide a template UEFI variable with `virtualisation.efi.variables` ; if this option is important to you, open an issue")
+    (mkRemovedOptionModule [ "virtualisation" "persistBootDevice" ] "Boot device is always persisted if you use a bootloader through the root disk image ; if this does not work for your usecase, please examine carefully what `virtualisation.{bootDevice, rootDevice, bootPartition}` options offers you and open an issue explaining your need.`")
```
Not really AFAIK, otherwise you get newlines (?). But maybe someone know how to achieve this.

RaitoBezarius commented on nixos/modules/virtualisation/qemu-vm.nix:
```diff
@@ -362,25 +324,48 @@ in
       };
 
-    virtualisation.bootDevice =
+    virtualisation.bootLoaderDevice =
```
> rootDrive for the main drive of the system, e.g. /dev/vda (when booting on UEFI this contains the ESP and the rootfs, when booting in legacy mode this only really contains the rootfs)

The problem with this naming is that it ignores that you can have a root drive that's different from your bootloader drive and what really matters is either:

- you have a bootloader and you boot on the bootloader device/drive/disk, be it a boot partition (ESP) or the 512 sectors for MBR
- you don't have it and bootLoaderDevice is absolutely irrelevant to you

Naming it rootDrive seems to make it relevant when you don't have a bootloader.

But really, what we want in bootloaderless mode is a "rootfs", kernel and initrd. Therefore, rootDrive is irrelevant.

RaitoBezarius replied:
So I don't really agree with your suggestion.

nikstur replied:
> Naming it rootDrive seems to make it relevant when you don't have a bootloader.

But you use the `bootLoaderDevice` even when you don't have a bootloader with: `"direct_boot_with_default_fs" = cfg.bootLoaderDevice;` in https://github.com/NixOS/nixpkgs/blob/07f5a705e2dd61e3f681640414c0e6f4c761aba1/nixos/modules/virtualisation/qemu-vm.nix#L261

Doesn't that make it relevant even when I don't have a bootloader? Or am I missing something?

RaitoBezarius replied:
I am open to change this naming in a further PR though :), I will resolve this for now (we can revisit it) and let's discuss this more.

RaitoBezarius commented on nixos/tests/bootspec.nix:
```diff
@@ -108,9 +108,9 @@ in
       machine.start()
       machine.wait_for_unit("multi-user.target")
 
-      machine.succeed("test -e /run/current-system/bootspec/boot.json")
```
I could do a PR for the sake of doing a PR but they're already in a separate atomic commit: https://github.com/NixOS/nixpkgs/pull/207039/commits/837efc6b31c62c4341df859f532a4df6188aa75f so I think it's harmless? I can do another PR if you want.

nikstur replied:
Not too worried about it. From my perspective it can stay in this PR.

RaitoBezarius commented on nixos/tests/bootspec.nix:
```diff
@@ -108,9 +108,9 @@ in
       machine.start()
       machine.wait_for_unit("multi-user.target")
 
-      machine.succeed("test -e /run/current-system/bootspec/boot.json")
```
Addressed in https://github.com/NixOS/nixpkgs/pull/207039/commits/837efc6b31c62c4341df859f532a4df6188aa75f

RaitoBezarius commented on nixos/modules/virtualisation/qemu-vm.nix:
```diff
@@ -113,6 +117,23 @@ let
       if test -n "$NIX_DISK_IMAGE" && ! test -e "$NIX_DISK_IMAGE"; then
           ${qemu}/bin/qemu-img create -f qcow2 "$NIX_DISK_IMAGE" \
             ${toString config.virtualisation.diskSize}M
```
Good point, I am doing this twice, also later. But we need to do a "copy on write" image backed by the system image to perform mutations as the system image is **read only**, a template, if you want.

RaitoBezarius replied:
(Remove the first useless creation.)

RaitoBezarius commented:
Started a new evaluation here: https://hydra.nixos.org/jobset/nixos/python-test-refactoring

RaitoBezarius commented:
Finally, this will be merged once ofborg CI passes.

RaitoBezarius commented:
Thanks to everyone involved in this journey. :)

alyssais commented:
This PR broke nixosTests.kexec.  It seems like the final Hydra eval had that test as broken, so maybe it's worth a look to see if any other tests were found to be broken during that eval and overlooked?

RaitoBezarius commented:
Hmm, I will investigate this further. I remember that some of the stuff
seemed already broken on master but I may have done mistakes. Thank you for
the heads-up!

Le ven. 19 mai 2023 à 23:45, Alyssa Ross ***@***.***> a
écrit :

> This PR broke nixosTests.kexec. It seems like the final Hydra eval had
> that test as broken, so maybe it's worth a look to see if any other tests
> were found to be broken during that eval and overlooked?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/207039#issuecomment-1555289200>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACMZRE4RFCHNMRJXUHLYU3XG7SZBANCNFSM6AAAAAATE4LUVY>
> .
> You are receiving this because you modified the open/close state.Message
> ID: ***@***.***>
>
